lib/ReplacePointerBitcastPass.cpp

```diff
@@ -208,838 +265,894 @@ Value *ConvertValue(Value *src, Type *dst_type, IRBuilder<> &builder) {
-                // ==> if types are same between src and dst, it will be
-                // igonored
-                //
-                unsigned SubNumElement = SrcEleTyBitWidth / DstTyBitWidth;
-                if (SubNumElement != 2 && SubNumElement != 4) {
-                  llvm_unreachable("Unsupported SubNumElement");
-                }
+  // If the output is a vec3 let's consider that the output is a vec4.
+  bool IsVec3 = DstTy->isVectorTy() && GetNumEle(DstTy) == 3;
 
-                Value *TmpOrgGEPIdx = Builder.CreateLShr(
-                    OrgGEPIdx, Builder.getInt32(std::log2(SubNumElement)));
-                Value *TmpVal = Builder.CreateExtractElement(
-                    LDValues[0], TmpOrgGEPIdx, "tmp_val");
-                TmpVal = Builder.CreateBitCast(
-                    TmpVal,
-                    FixedVectorType::get(
-                        IntegerType::get(DstTy->getContext(), DstTyBitWidth),
-                        SubNumElement));
-                TmpOrgGEPIdx = Builder.CreateAnd(
-                    OrgGEPIdx, Builder.getInt32(SubNumElement - 1));
-                TmpVal = Builder.CreateExtractElement(TmpVal, TmpOrgGEPIdx,
-                                                      "tmp_val");
-                DstVal = Builder.CreateBitCast(TmpVal, DstTy);
-              } else {
-                Inst->print(errs());
-                llvm_unreachable("Handle this bitcast");
-              }
-            }
-          } else if (SrcTyBitWidth < DstTyBitWidth) {
-            //
-            // Consider below case.
-            //
-            // Original IR (float2* --> float4*)
-            // 1. src_addr = bitcast float2*, float4*
-            // 2. element_addr = gep (float4*) src_addr, idx
-            // 3. load (float4*) element_addr
-            //
-            // Transformed IR
-            // 1. src_addr = gep (float2*) src, idx * 2
-            // 2. src_val1(float2) = load (float2*) src_addr
-            // 3. src_addr2 = gep (float2*) src_addr, 1
-            // 4. src_val2(float2) = load (float2*) src_addr2
-            // 5. dst_val(float4) = shufflevector src_val1, src_val2, <0, 1>
-            // 6. dst_val(float4) = bitcast dst_val, (float4)
-            // ==> if types are same between src and dst, it will be igonored
-            //
-            unsigned NumElement = 1;
-            if (SrcTy->isVectorTy()) {
-              NumElement = SrcVecTy->getElementCount().getKnownMinValue() * 2;
-            }
+  // Because the vec3 to vec4 pass is before this one, we should not have a vec3
+  // src. But it seems that some llvm passes after vec3 to vec4 can produce new
+  // vec3. At the moment the only case known is to produce vec3 that will be
+  // bitcast to another vec3 which element has the same time as the src vec3. In
+  // that particular case, just keep the vec3 as we only need to bitcast them,
+  // which will be handled correctly by this pass.
```

Comment:
This is not clear. If the only known case of the source being a vec3 is when casting to another vec3 then wouldn't `IsVec3` already be true? Maybe this should be an assert?

Reply:
This condition `IsVec3` is used to avoid expecting a vec3 output from `combineInto`. Instead, we expect a vec4 output, and then we convert the vec4 to vec3.
When llvm creates this vec3 to vec3 bitcast with elements of the same size, we do not want/need to expect a vec4 output from `combineInto`. We can just ask `combineInto` to produce a vec3.

This is just to have a more efficient and understandable  output from this pass. Removing this condition would work as well, but it would produce unneeded transformation that is hard to understand when you just look at the test concerned by it.